# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/mediasink/AudioSinkWrapper.h
# Commit: 048e93d0a274
# Full Hash: 048e93d0a274fab7d58e5b601cbe55b911ae9c00
# Author: Paul Adenot <paul@paul.cx>
# Date: 2022-04-20 21:53:00
# Regressor Bug: 1743834
# File Overlap Count: 1
# Description:
#   Bug 1743834 - Shut down the AudioSink when audio is muted. r=alwu,media-playback-reviewers
#   
#   This does the following:
#   - When the media is muted, shut down and release the AudioSink ;
#   - While the media is muted, use the system clock to make video advance ;
# ==============================================================================

diff -r 32210507c793 -r 048e93d0a274 dom/media/mediasink/AudioSinkWrapper.h
--- a/dom/media/mediasink/AudioSinkWrapper.h	Wed Apr 20 12:07:09 2022 +0000
+++ b/dom/media/mediasink/AudioSinkWrapper.h	Wed Apr 20 12:07:09 2022 +0000
@@ -66,6 +66,7 @@
   media::TimeUnit GetPosition(TimeStamp* aTimeStamp = nullptr) override;
   bool HasUnplayedFrames(TrackType aType) const override;
   media::TimeUnit UnplayedDuration(TrackType aType) const override;
+  void DropAudioPacketsIfNeeded(const media::TimeUnit& aMediaPosition);
 
   void SetVolume(double aVolume) override;
   void SetStreamName(const nsAString& aStreamName) override;
@@ -86,6 +87,8 @@
   void GetDebugInfo(dom::MediaSinkDebugInfo& aInfo) override;
 
  private:
+  bool IsMuted() const;
+  void OnMuted(bool aMuted);
   virtual ~AudioSinkWrapper();
 
   void AssertOwnerThread() const {
@@ -94,7 +97,11 @@
 
   nsresult StartAudioSink(const media::TimeUnit& aStartTime);
 
-  media::TimeUnit GetVideoPosition(TimeStamp aNow) const;
+  // Get the current media position using the system clock. This is used when
+  // the audio is muted, or when the media has no audio track. Otherwise, the
+  // media's position is based on the clock of the AudioStream.
+  media::TimeUnit GetSystemClockPosition(TimeStamp aNow) const;
+  bool CheckIfEnded() const;
 
   void OnAudioEnded();
 
